src/lib/Libifl/pbsD_connect.c

```diff
@@ -470,12 +349,186 @@ __pbs_connect_extend(char *server, char *extend_data)
+ *
+ * @return	void
+ * @retval	!NULL - success
+ * @retval	NULL - error
+ *
+ * @par Side Effects:
+ *	None
+ *
+ * @par MT-safe: Yes
+ */
+void *
+get_conn_servers(void)
+{
+	svr_conn_t *conn_arr = NULL;
+
+	conn_arr = pthread_getspecific(psi_key);
+	if (conn_arr == NULL && pbs_conf.psi != NULL) {
+		int num_svrs;
+		int i;
+
+		num_svrs = get_num_servers();
+		conn_arr = calloc(num_svrs, sizeof(svr_conn_t));
+		if (conn_arr == NULL) {
+			pbs_errno = PBSE_SYSTEM;
+			return NULL;
+		}
+
+		for (i = 0; i < num_svrs; i++) {
+			strcpy(conn_arr[i].name, pbs_conf.psi[i].name);
+			conn_arr[i].port = pbs_conf.psi[i].port;
+			conn_arr[i].sd = -1;
+			conn_arr[i].secondary_sd = -1;
+			conn_arr[i].state = SVR_CONN_STATE_DOWN;
+		}
+
+		pthread_setspecific(psi_key, conn_arr);
+	}
+
+	return conn_arr;
+}
+
+
+/**
+ * @brief	Helper function for connect_to_servers to connect to a particular server
+ *
+ * @param[in]		idx - array index for the server to connect to
+ * @param[in,out]	conn_arr - array of svr_conn_t
+ * @param[in]		extend_data - any additional data relevant for connection
+ *
+ * @return	int
+ * @retval	-1 for error
+ * @retval	fd of connection
+ */
+static int
+connect_to_server(int idx, svr_conn_t *conn_arr, char *extend_data)
+{
+	if (conn_arr[idx].state != SVR_CONN_STATE_UP) {
+		if ((conn_arr[idx].sd = tcp_connect(conn_arr[idx].name, conn_arr[idx].port, extend_data)) != -1) {
+			conn_arr[idx].state = SVR_CONN_STATE_UP;
+			pbs_client_thread_lock_conntable();
+			add_connection(conn_arr[idx].sd);
+			pbs_client_thread_unlock_conntable();
+		}
+		else
+			conn_arr[idx].state = SVR_CONN_STATE_DOWN;
+	}
+
+	return conn_arr[idx].sd;
+}
+
+/**
+ * @brief	To connect to all the servers
+ *
+ * @param[in]	server_name - name of the server to connect to (NULL if not known)
+ * @param[in]	port - port of the server to connect to (considered if server_name is not NULL)
+ * @param[in]	extend_data
+ *
+ * @return int
+ * @retval >0 - success
+ * @retval -1 - error
+ */
+static int
+connect_to_servers(char *server_name, uint port, char *extend_data)
+{
+	int i;
+	int fd = -1;
+	int ret_fd = -1;
+	int multi_flag = 0;
+	int num_conf_servers = get_num_servers();
+	svr_conn_t *svr_connections = get_conn_servers();
+
+	if (svr_connections == NULL)
+		return -1;
+
+	multi_flag = getenv(CONN_MULTI) != NULL;
+	if (!multi_flag) {	/* Client wants to connect to only 1 server */
+		if (server_name) {
+			for (i = 0; i < num_conf_servers; i++) {
+				if (!strcmp(server_name, pbs_conf.psi[i].name) && port == pbs_conf.psi[i].port) {
+					/* Client wants to connect to a specific server, so we return here regardless of result */
+					return connect_to_server(i, svr_connections, extend_data);
+				}
+			}
+		}
+
+		/* Client doesn't care which server to connect to, try random first */
+		i = rand_num() % num_conf_servers;
+		fd = connect_to_server(i, svr_connections, extend_data);
+		if (fd != -1)
+			return fd;
+
+		/* didn't work, we'll try all servers below */
```

Comment:
This comment is misleading. In the absence of multi_flag we only connect to one of the random servers which is active. Can you please remove the comment or rephrase it.